core/src/main/java/org/elasticsearch/common/xcontent/XContentType.java

```diff
@@ -108,31 +113,26 @@ public String shortName() {
         public XContent xContent() {
             return CborXContent.cborXContent;
         }
-    },;
+    };
 
-    public static XContentType fromRestContentType(String contentType) {
-        if (contentType == null) {
+    public static XContentType fromMediaTypeOrFormat(String mediaType) {
+        if (mediaType == null) {
             return null;
         }
-        if ("application/json".equals(contentType) || "json".equalsIgnoreCase(contentType)) {
-            return JSON;
-        }
-
-        if ("application/smile".equals(contentType) || "smile".equalsIgnoreCase(contentType)) {
-            return SMILE;
-        }
-
-        if ("application/yaml".equals(contentType) || "yaml".equalsIgnoreCase(contentType)) {
-            return YAML;
-        }
-
-        if ("application/cbor".equals(contentType) || "cbor".equalsIgnoreCase(contentType)) {
-            return CBOR;
+        for (XContentType type : values()) {
+            if (isMediaType(mediaType, type)) {
+                return type;
+            }
         }
-
         return null;
     }
 
+    private static boolean isMediaType(String stringType, XContentType type) {
+        return type.mediaTypeName().equals(stringType) ||
+                stringType.startsWith(type.mediaTypeName() + ";") ||
```

Comment:
This comparison should also be a [case-insensitive comparison](http://tools.ietf.org/html/rfc2045#section-2), and should take my [previous comment](https://github.com/elastic/elasticsearch/pull/14421/files#r44447472) into consideration.